### PR TITLE
Fix #1206 by passing instance name as anchor

### LINKED ISSUE #1206: Can't un-collapse collapsible example on instances
Take a look at the documentation on the `Semigroup` and `Monoid` instances here:

http://hackage.haskell.org/package/unordered-containers-0.2.11.0/docs/Data-HashMap-Strict.html#t:HashMap

If I try to un-collapse / view the example for the `Semigroup` instance, the example for the `Monoid` instance is revealed instead.

Clicking on the `Monoid` example also reveals the `Monoid` example, so the `Semigroup` example seems impossible to view.

image

---

**EDIT:** I can reproduce this both with Chromium and Firefox.

### PR DESCRIPTION
**Pull request by soupi**
Hi! Trying my hand at contributing.

tl;dr: I added the instance name (+ number) as an anchor instead of the current default anchor for collapsing examples in instances. It might be worth it to replace the `Maybe` anchor to just an anchor to avoid situations like this.

I tracked this bug to be originated from [hackMarkup](https://github.com/haskell/haddock/blob/ghc-8.10/haddock-api/src/Haddock/Backends/Xhtml/DocMarkup.hs#L184), The `nm` in `CollapsingHeader` does not receive a meaningful name from (up the chain) `subRow` and therefore defaults to `noid:`.

I looked for the latest place I can still find the instance name and added it as another parameter. Is it possible to extract the instance name somewhere else? I couldn't figure it out.

Also, is there a different name I should use instead of `String`? Actually I don't think generating an anchor name is the responsibility of `subInstances` etc, but I couldn't figure out where is a good place to do this while still having the information I needed.

Also, where is a good way add an automated test for this property (unique anchor names)?

Thanks!

### CODE REVIEW

**Comment by sjakobi** [comment]
Thanks for working on this @soupi! :) 

I think a regression test in the `html-test` suite would be nice to have. See https://github.com/haskell/haddock#updating-golden-testsuite-outputs.

**Comment by soupi** [comment]
Thanks for the tip @sjakobi, I've added a regression test to html-test.

File: haddock-api/src/Haddock/Backends/Xhtml/Decl.hs [hunk 1]
```diff
@@ -634,9 +634,9 @@ ppInstances links origin instances splice unicode pkg qual
   -- force Splice = True to use line URLs
   where
     instName = getOccString origin
-    instDecl :: Int -> DocInstance DocNameI -> (SubDecl, Maybe Module, Located DocName)
+    instDecl :: Int -> DocInstance DocNameI -> (String, SubDecl, Maybe Module, Located DocName)
     instDecl no (inst, mdoc, loc, mdl) =
-        ((ppInstHead links splice unicode qual mdoc origin False no inst mdl), mdl, loc)
+        (getOccString (ihdClsName inst) <> "_" <> show no <> ":", (ppInstHead links splice unicode qual mdoc origin False no inst mdl), mdl, loc)
```
Comment (sjakobi) [inline haddock-api/src/Haddock/Backends/Xhtml/Decl.hs:639]: I think defining the added `String` and the `Maybe Module` in separate in a where-clause would help with readability.

File: haddock-api/src/Haddock/Backends/Xhtml/Decl.hs [hunk 2]
```diff
@@ -649,9 +649,9 @@ ppOrphanInstances links instances splice unicode pkg qual
     instOrigin :: InstHead name -> InstOrigin (IdP name)
     instOrigin inst = OriginClass (ihdClsName inst)
 
-    instDecl :: Int -> DocInstance DocNameI -> (SubDecl, Maybe Module, Located DocName)
+    instDecl :: Int -> DocInstance DocNameI -> (String, SubDecl, Maybe Module, Located DocName)
     instDecl no (inst, mdoc, loc, mdl) =
-        ((ppInstHead links splice unicode qual mdoc (instOrigin inst) True no inst Nothing), mdl, loc)
+        (getOccString (ihdClsName inst) <> "_" <> show no <> ":", (ppInstHead links splice unicode qual mdoc (instOrigin inst) True no inst Nothing), mdl, loc)
```
Comment (sjakobi) [inline haddock-api/src/Haddock/Backends/Xhtml/Decl.hs:654]: Same situation here.

**Review by sjakobi — APPROVED** [review]
Two small wibbles. Otherwise LGTM! :+1:

**Comment by soupi** [comment]
Thanks for the review @sjakobi :+1: I've made the requested changes.

**Review by alexbiehl — APPROVED** [review]
This is looking good. Let's get this in. Please also cherry-pick this PR on top of ghc-head branch.

**Comment by Kleidukos** [comment]
@soupi Is there any chance you can re-open a PR from ghc-9.0? :)
